src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -1686,77 +1672,60 @@ private void MakeSeparatorList(ReadOnlySpan<char> separators, ref ValueListBuild
-
-            ref char c0 = ref MemoryMarshal.GetReference(this.AsSpan());
-            int cond = Length & -Vector128<ushort>.Count;
             int i = 0;
-
-            for (; i < cond; i += Vector128<ushort>.Count)
+            if (Vector256.IsHardwareAccelerated && Length >= Vector256<ushort>.Count * 2)
             {
-                Vector128<ushort> charVector = ReadVector(ref c0, i);
-                Vector128<ushort> cmp = Sse2.CompareEqual(charVector, v1);
-
-                cmp = Sse2.Or(Sse2.CompareEqual(charVector, v2), cmp);
-                cmp = Sse2.Or(Sse2.CompareEqual(charVector, v3), cmp);
-
-                if (Sse41.TestZ(cmp, cmp)) { continue; }
-
-                Vector128<byte> mask = Sse2.ShiftRightLogical(cmp.AsUInt64(), 4).AsByte();
-                mask = Ssse3.Shuffle(mask, shuffleConstant);
+                ref ushort source = ref Unsafe.As<char, ushort>(ref _firstChar);
 
-                uint lowBits = Sse2.ConvertToUInt32(mask.AsUInt32());
-                mask = Sse2.ShiftRightLogical(mask.AsUInt64(), 32).AsByte();
-                uint highBits = Sse2.ConvertToUInt32(mask.AsUInt32());
+                Vector256<ushort> v1 = Vector256.Create((ushort)c);
+                Vector256<ushort> v2 = Vector256.Create((ushort)c2);
+                Vector256<ushort> v3 = Vector256.Create((ushort)c3);
 
-                for (int idx = i; lowBits != 0; idx++)
+                int vector256ShortCount = Vector256<ushort>.Count;
+                for (; (i + vector256ShortCount) <= Length; i += vector256ShortCount)
                 {
-                    if ((lowBits & 0xF) != 0)
+                    Vector256<ushort> vector = Vector256.LoadUnsafe(ref source, (uint)i);
+                    Vector256<ushort> cmp = Vector256.Equals(vector, v1) | Vector256.Equals(vector, v2) | Vector256.Equals(vector, v3);
+
+                    uint mask = cmp.AsByte().ExtractMostSignificantBits() & 0b0101010101010101;
+                    while (mask != 0)
                     {
-                        sepListBuilder.Append(idx);
+                        sepListBuilder.Append(i + BitOperations.TrailingZeroCount(mask) / 2);
+                        mask = BitOperations.ResetLowestSetBit(mask);
                     }
-
-                    lowBits >>= 8;
                 }
+            }
+            else if (Vector128.IsHardwareAccelerated && Length >= Vector128<ushort>.Count * 2)
+            {
+                ref ushort source = ref Unsafe.As<char, ushort>(ref _firstChar);
+
+                Vector128<ushort> v1 = Vector128.Create((ushort)c);
+                Vector128<ushort> v2 = Vector128.Create((ushort)c2);
+                Vector128<ushort> v3 = Vector128.Create((ushort)c3);
 
-                for (int idx = i + 4; highBits != 0; idx++)
+                int vector128ShortCount = Vector128<ushort>.Count;
+                for (; (i + vector128ShortCount) <= Length; i += vector128ShortCount)
```

Comment:
```suggestion
                for (; i <= Length - Vector128<ushort>.Count; i += Vector128<ushort>.Count)
```
When `i` is of type `nint` just check if the comparison doesn't introduce any sign extensions -- please double check to be on the safe side.